### PR TITLE
feat(core): dual-pressure shouldRespond, docs, and API surface

### PR DESCRIPTION
**Pull request by odilitime**
Implement speak_up/hold_back scoring with net-based clamp (REPLY/RESPOND to IGNORE when net <= -threshold), STOP exemption, and soft warn on high net + IGNORE. Add DUAL_PRESSURE_THRESHOLD env, MessageProcessingResult fields, and optional ResponseDecision pressure/classifierAction. Drop ANXIETY from initial composeState for the message pipeline; sync basic-capabilities shouldRespond return type and bypass env aliases.

Document behavior in SHOULD_RESPOND_DUAL_PRESSURE.md, DESIGN.md, ROADMAP.md, CHANGELOG, and READMEs; expand JSDoc and prompt template comments. Tests cover clamp, warn, result fields, and prompt ordering.

Made-with: Cursor

---

> [!NOTE]
> **Medium Risk**
> Changes the core message gating logic and prompt schema for ambiguous channels, which can materially affect when agents reply or stay silent. Risk is mitigated by a configurable threshold, STOP exemption, and added tests/docs, but behavioral regressions are still possible across deployments and custom templates.
> 
> **Overview**
> Introduces a **dual-pressure** LLM `shouldRespond` classifier for ambiguous channels that outputs `speak_up`/`hold_back` (0–100) plus `action`, computes `net`, and applies runtime consistency rules: **clamp** REPLY/RESPOND to IGNORE when `net <= -DUAL_PRESSURE_THRESHOLD`, **never** override STOP, and **warn** on high-net IGNORE.
> 
> Extends the public API surface by adding `DualPressureScores`, exposing `dualPressure` and `shouldRespondClassifierAction` on `MessageProcessingResult`, and aligning `basic-capabilities` `shouldRespond` to return `ResponseDecision` with env alias fallbacks. Updates the default `shouldRespondTemplate`, removes `ANXIETY` from the initial `composeState` provider list for this pipeline, adds `DUAL_PRESSURE_THRESHOLD` to `.env.example`, and adds focused tests plus new documentation (`docs/SHOULD_RESPOND_DUAL_PRESSURE.md`, `ROADMAP.md`, DESIGN/README links).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 096a60b162b08e362ecfe2f490219cc812d75f3b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Dual-pressure scoring (speak_up / hold_back) for should-respond decisions and a DUAL_PRESSURE_THRESHOLD setting to control clamping behavior
  * Response metadata now surfaces dual-pressure scores and classifier action details

* **Documentation**
  * New dual-pressure design doc, updated READMEs, changelog, and roadmap entries describing behavior and configuration

* **Tests**
  * Added tests covering clamping, warning, and template/schema changes for the classifier

* **Refactor**
  * Updated response decision/type surfaces to include optional pressure/classifier fields

<h3>Greptile Summary</h3>

This PR introduces a dual-pressure `shouldRespond` classifier that replaces the single-action LLM call with explicit `speak_up` / `hold_back` scores (0–100) plus an `action` field. A runtime clamp ("gaslighting guard") overrides REPLY/RESPOND to IGNORE when `net ≤ -DUAL_PRESSURE_THRESHOLD`, while STOP is always preserved. New fields on `MessageProcessingResult` expose classifier outcomes for tests and dashboards. Additional changes: `ANXIETY` dropped from initial `composeState`, env aliases for bypass lists, and updated JSDoc/docs.

All remaining findings are P2 (style/observability); no blocking correctness issues were found.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all findings are P2 style/observability issues that do not affect correctness in the default configuration.

The core clamp/warn logic is correctly implemented and covered by new tests. Boundary conditions (STOP exemption, partial-score survival, empty-action fallback) are handled. The P2 items (T_hi mismatch when threshold is overridden, empty-string sentinel leaking into results, misleading STOP example) do not affect runtime behavior at the default DUAL_PRESSURE_THRESHOLD=20 and do not constitute data loss or silent misclassification in the expected operating range.

packages/typescript/src/prompts.ts (T_hi hardcoding) and packages/typescript/src/services/message.ts (empty-string sentinel).

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/typescript/src/services/message.ts | Core logic: adds `applyDualPressureToClassifierAction` with clamp/warn and wires `dualPressure`/`shouldRespondClassifierAction` into the result; T_hi mismatch and empty-string sentinel are notable edge cases. |
| packages/typescript/src/prompts.ts | Prompt rewritten to dual-pressure format with `speak_up`/`hold_back` fields; `T_hi = 20` is hardcoded and won't reflect operator overrides to `DUAL_PRESSURE_THRESHOLD`. |
| packages/typescript/src/types/message-service.ts | Adds `DualPressureScores` interface and extends `MessageProcessingResult` / `ResponseDecision` with optional pressure fields; types are correct and well-documented. |
| packages/typescript/src/basic-capabilities/index.ts | Return type aligned to `ResponseDecision`; env alias fallback via `||` for bypass list settings is straightforward and safe. |
| packages/typescript/src/__tests__/message-service.test.ts | New tests cover clamp-to-IGNORE and high-net-IGNORE-warn paths; assertions on `dualPressure` object and `shouldRespondClassifierAction` are correct. |
| packages/typescript/src/__tests__/prompts.test.ts | Updated to assert new `speak_up`/`hold_back` fields, `dual_pressure[2]`, `anti_gaming`, and ordering of `action_space` before `output`. |
| .env.example | Documents `DUAL_PRESSURE_THRESHOLD` with correct default, range, and semantics. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Incoming Message] --> B{isAutonomous?}
    B -- yes --> F[shouldRespondToMessage = true]
    B -- no --> C{shouldRespond rule-based}
    C -- DM/API/bypass skipEvaluation=true --> F
    C -- ambiguous GROUP channel --> D[LLM Classifier\ndynamicPromptExecFromState]
    D --> E[applyDualPressureToClassifierAction]
    E --> E1{scores parseable?}
    E1 -- no / partial --> E2[pressure = null\nfinalActionUpper = rawAction]
    E1 -- yes --> E3{action = STOP?}
    E3 -- yes --> E4[return STOP\nnever clamped]
    E3 -- no --> E5{net <= -threshold\nAND isEngage?}
    E5 -- yes --> E6[warn: clamping\nfinalActionUpper = IGNORE]
    E5 -- no --> E7{net >= threshold\nAND action=IGNORE?}
    E7 -- yes --> E8[warn: high net but IGNORE\nallow model decision]
    E7 -- no --> E9[pass through action]
    E2 --> G
    E4 --> G
    E6 --> G
    E8 --> G
    E9 --> G
    G{finalActionUpper\nin nonResponseActions?} -- IGNORE / STOP / NONE --> H[shouldRespondToMessage = false]
    G -- REPLY / RESPOND / other --> F
    H --> I[Return result\ndualPressure + shouldRespondClassifierAction]
    F --> J[Run message pipeline]
    J --> I
```

<sub>Reviews (1): Last reviewed commit: ["feat(core): dual-pressure shouldRespond,..."](https://github.com/elizaos/eliza/commit/484c7060b26442d5878ea62e8ec94885d538d1bb) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=27963751)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

**Comment by odilitime** [comment]
@coderabbitai review

File: packages/typescript/src/services/message.ts [hunk 1]
```diff
@@ -1218,20 +1363,38 @@ export class DefaultMessageService implements IMessageService {
 
 				// A classifier output can either continue the turn or terminate it.
 				const nonResponseActions = ["IGNORE", "NONE", "STOP"];
-				const actionValue = responseObject?.action;
+				const rawAction =
+					typeof responseObject?.action === "string"
+						? responseObject.action
+						: "";
+				const dual = applyDualPressureToClassifierAction(
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/services/message.ts]: Fixed in `4be4174`.

File: packages/typescript/src/prompts.ts [hunk 2]
```diff
@@ -627,23 +652,42 @@ decision_note:
...
+
+Example:
+name: {{agentName}}
+speak_up: 40
+hold_back: 35
+reasoning: User explicitly asked me to stop talking.
+action: STOP
+primaryContext: general
+secondaryContexts:
 evidenceTurnIds:`;
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/prompts.ts:692]: Fixed in `4be4174`.

File: packages/typescript/src/prompts.ts [hunk 3]
```diff
@@ -602,21 +602,46 @@ decision: true`;
...
+dual_pressure[2]:
+- speak_up: integer 0-100 — pressure TO engage (direct address, question in domain, obligation, unique value, user need; group-wide address; topic fit)
+- hold_back: integer 0-100 — pressure to STAY QUIET (wrong audience, already answered, redundancy, noisy channel, better-qualified others, low unique value, social norms)
+
+net: speak_up minus hold_back (range -100 to +100).
+
+consistency (T_hi = 20 unless you state otherwise in reasoning):
+- net >= +T_hi -> prefer REPLY over IGNORE unless you document an exception in reasoning
+- net <= -T_hi -> prefer IGNORE over REPLY (or STOP if the user asked to stop)
+- between -T_hi and +T_hi -> judgment; explain tradeoffs in reasoning
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/prompts.ts]: Fixed in `4be4174`.

File: packages/typescript/src/services/message.ts [hunk 4]
```diff
@@ -1218,20 +1363,38 @@ export class DefaultMessageService implements IMessageService {
...
+					runtime,
+					responseObject as Record<string, unknown> | null,
+					rawAction,
+				);
+				dualPressureLog = dual.pressure;
+				const actionUpper = dual.finalActionUpper;
+				shouldRespondClassifierAction = actionUpper;
+
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/services/message.ts]: Fixed in `4be4174`.

File: packages/typescript/src/basic-capabilities/index.ts [hunk 5]
```diff
@@ -388,10 +389,12 @@ export function shouldRespond(
...
 	const customChannels = normalizeEnvList(
-		runtime.getSetting("ALWAYS_RESPOND_CHANNELS"),
+		runtime.getSetting("ALWAYS_RESPOND_CHANNELS") ||
+			runtime.getSetting("SHOULD_RESPOND_BYPASS_TYPES"),
 	);
 	const customSources = normalizeEnvList(
-		runtime.getSetting("ALWAYS_RESPOND_SOURCES"),
+		runtime.getSetting("ALWAYS_RESPOND_SOURCES") ||
+			runtime.getSetting("SHOULD_RESPOND_BYPASS_SOURCES"),
 	);
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/basic-capabilities/index.ts:407]: Fixed in `4be4174`.

File: packages/typescript/src/basic-capabilities/index.ts [hunk 6]
```diff
@@ -388,10 +389,12 @@ export function shouldRespond(
 	const alwaysRespondSources = ["client_chat"];
 
 	const customChannels = normalizeEnvList(
-		runtime.getSetting("ALWAYS_RESPOND_CHANNELS"),
+		runtime.getSetting("ALWAYS_RESPOND_CHANNELS") ||
+			runtime.getSetting("SHOULD_RESPOND_BYPASS_TYPES"),
 	);
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/basic-capabilities/index.ts:401]: Fixed in `4be4174`.

File: packages/typescript/src/prompts.ts [hunk 7]
```diff
@@ -627,23 +652,42 @@ decision_note:
 - talking TO {{agentName}} means name mention, reply chain, or direct continuation
 - talking ABOUT {{agentName}} is not enough
 
+action_space:
+- REPLY: full conversational response is warranted
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/prompts.ts:656]: Fixed in `4be4174`.

File: packages/typescript/src/prompts.ts [hunk 8]
```diff
@@ -602,21 +602,46 @@ decision: true`;
...
 export const SHOULD_MUTE_ROOM_TEMPLATE = shouldMuteRoomTemplate;
 
-export const shouldRespondTemplate = `task: Decide whether {{agentName}} should respond, ignore, or stop.
+/**
+ * Default should-respond classifier prompt (TOON). Uses dual scores (`speak_up`, `hold_back`) so the model
+ * must argue both sides before `action`. **Why:** reduces over-reply and action/score mismatch; runtime
+ * clamp/warn enforces consistency with `DUAL_PRESSURE_THRESHOLD`. `action_space` precedes `output:` so
+ * labels are defined before the format constraint. See `docs/SHOULD_RESPOND_DUAL_PRESSURE.md`.
+ */
+export const shouldRespondTemplate = `task: Decide whether {{agentName}} should REPLY, IGNORE, or STOP, using dual-pressure scoring.
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/prompts.ts]: Fixed in `4be4174`.

File: packages/typescript/src/services/message.ts [hunk 9]
```diff
@@ -1630,6 +1795,13 @@ export class DefaultMessageService implements IMessageService {
 			responseMessages,
 			state,
 			mode,
+			// Expose classifier outcome only when the LLM path ran (null sentinel set in that branch).
+			...(shouldRespondClassifierAction !== null
+				? {
+						dualPressure: dualPressureLog,
+						shouldRespondClassifierAction,
+					}
+				: {}),
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/services/message.ts]: Fixed in `4be4174`.

File: packages/typescript/src/__tests__/message-service.test.ts [hunk 10]
```diff
@@ -138,6 +138,17 @@ describe("DefaultMessageService", () => {
...
+			name: runtime.character.name ?? "TestAgent",
+			speak_up: 72,
+			hold_back: 18,
+			reasoning: "Default test dual-pressure decision",
+			action: "RESPOND",
+			primaryContext: "general",
+			secondaryContexts: "",
+			evidenceTurnIds: "",
+		});
+
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/__tests__/message-service.test.ts]: Fixed in `4be4174`.

File: packages/typescript/src/prompts.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/prompts.ts]: Fixed in `4be4174`.

File: packages/typescript/src/__tests__/prompts.test.ts [hunk 11]
```diff
@@ -17,19 +17,26 @@ describe("Prompts", () => {
...
 			expect(shouldRespondTemplate).toContain(
 				"direct mention of {{agentName}}",
 			);
 			expect(shouldRespondTemplate).toContain("decision_note:");
 			expect(shouldRespondTemplate).toContain("talking TO {{agentName}}");
+			expect(shouldRespondTemplate).toContain("dual_pressure[2]:");
+			expect(shouldRespondTemplate).toContain("anti_gaming:");
+			expect(shouldRespondTemplate.indexOf("action_space:")).toBeLessThan(
+				shouldRespondTemplate.indexOf("output:"),
+			);
```
[the remark this answers is not in the document]
Comment (odilitime) [reply packages/typescript/src/__tests__/prompts.test.ts]: Fixed in `4be4174`.

**Comment by odilitime** [comment]
@coderabbitai review

**Comment by odilitime** [comment]
@coderabbitai review

**Comment by odilitime** [comment]
@coderabbitai review

**Comment by lalalune** [comment]
Closing — Cursor-generated core shouldRespond changes. The dual-pressure concept is interesting but needs to be developed more carefully before landing in the message processing hot path. Consider proposing this as a standalone plugin that hooks into shouldRespond rather than modifying core.
